yogstation/code/datums/components/lingering.dm

```diff
@@ -26,53 +38,69 @@
 		START_PROCESSING(SSobj, src)
 
 /datum/component/lingering/proc/Entered(datum/source, atom/movable/AM)
-	if(affect_stuff(AM))
+	if(affect_target(AM))
 		START_PROCESSING(SSobj, src)
 
 /datum/component/lingering/process(delta_time)
-	if(!affect_stuff(null, delta_time))
+	if(!affect_stuff(delta_time))
 		STOP_PROCESSING(SSobj, src)
 
 ////////////////////////////////////////////////////////////////////////////////////
 //---------------------------------safety check-----------------------------------//
 ////////////////////////////////////////////////////////////////////////////////////
 /datum/component/lingering/proc/is_safe()
-	//if anything matching this typecache is found in the lava, we don't burn things
-	var/static/list/lava_safeties_typecache = typecacheof(list(/obj/structure/lattice/catwalk, /obj/structure/stone_tile))
+	if(!safeties_typecache)
+		return FALSE
 
 	var/turf/open/place = get_turf(parent)
 
-	var/list/found_safeties = typecache_filter_list(place.contents, lava_safeties_typecache)
+	var/list/found_safeties = typecache_filter_list(place.contents, safeties_typecache)
 
-	for(var/obj/structure/stone_tile/S in found_safeties)
+	for(var/obj/structure/stone_tile/S in found_safeties) //snowflake check because stone tiles work weirdly, not ideal
 		if(S.fallen)
 			LAZYREMOVE(found_safeties, S)
+
 	return LAZYLEN(found_safeties)
 
-/datum/component/lingering/proc/affect_stuff(AM, delta_time = 1)
-	. = 0
+////////////////////////////////////////////////////////////////////////////////////
+//-----------------------------single tick upon enter-----------------------------//
+////////////////////////////////////////////////////////////////////////////////////
+/datum/component/lingering/proc/affect_target(AM)
+	. = FALSE
 
 	if(!affect_callback)
 		stack_trace("Lingering component has been applied to a turf without an affect callback proc") //covering my bases
 		return
 
-	if(is_safe())
-		return FALSE
+	if(is_safe()) //we don't want to damage the target, but we want to start processing just in case
+		return TRUE
 
 	if(AM) //if it's a specific object, apply to that one
-		return affect_callback.Invoke(AM, delta_time)
+		return affect_callback.Invoke(AM, 1)
+
+////////////////////////////////////////////////////////////////////////////////////
+//--------------------------Tick everything inside on process---------------------//
+////////////////////////////////////////////////////////////////////////////////////
+/datum/component/lingering/proc/affect_stuff(delta_time = 1)
+	. = FALSE
+
+	if(!affect_callback)
+		stack_trace("Lingering component has been applied to a turf without an affect callback proc") //covering my bases
+		return
 
 	var/turf/open/place = get_turf(parent) //otherwise, apply to every object on the turf
 
 	if(!place)
 		return
 
 	for(var/thing in place.contents)
+		if(is_safe()) //we don't want to damage anything, but if there's something here, make sure to keep processing just in case it stops being safe
```

Comment:
Unless I'm missing something, the safety check should be outside the loop to prevent calling it multiple times needlessly

Reply:
ah, true, i want it to return in the loop so it only returns if theres something there, but no reason to call is_safe multiple times